IDE0300 fix-all skips first diagnostic in a cascading ternary

**Version Used**: 
Version 17.9.0 Preview 3.0 [34431.11.main]

**Steps to Reproduce**:
```C#
int value = 3;
M(value is 1 ? new[] { "1" } :
  value is 2 ? new[] { "2" } :
  value is 3 ? new[] { "3" } :
  new[] { "4" });

static void M(string[] args) { }
```

**Diagnostic Id**:
IDE0300

**Expected Behavior**:
Four IDE0300 diagnostics issued. Selecting the last one and choosing to fix-all fixes all of them:
```C#
int value = 3;
M(value is 1 ? ["1"] :
  value is 2 ? ["2"] :
  value is 3 ? ["3"] :
  ["4"]);

static void M(string[] args) { }
```

**Actual Behavior**:
Four IDE0300 diagnostics issued. Selecting the last one and choosing to fix-all fixes the last three but not the first:
```C#
int value = 3;
M(value is 1 ? new[] { "1" } :
  value is 2 ? ["2"] :
  value is 3 ? ["3"] :
  ["4"]);

static void M(string[] args) { }
```
cc: @CyrusNajmabadi

Also, I think there are actual null safety issues which have the same root cause. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzggPgYhgVwDZ4TB4CmABCTEaQLABQAAgMxkMBMZAwmQN71kDWLBgEYADAG0AumQCyIgBTAA9srxlgASl79BegG4QATmWMmAvBrIB+MhIBEI+zJBkYJAO7TeZe23tkAL4A3Lp6AmYSYjKWuATBZAD0iWRgUCgosADmZB7GMNlh4QwA7KZGRqF0eoH0Rcys4t6ybEqq6lo61eEChiZmZLGe3jy+/kFVPREVg1a2Dk4u5ZVFepHRs3F4Ccm5+dlcAMoAHABsbACsrpwQMDDKCGQAxsow+iRGj1tkeFAIHxB1AhlG5XgBaLbUchGEgAMw+lCe5AQAE8AA4kAB0q0EpWWk0EtTogSAA)

```cs
#nullable enable
public class C {
    public string[] M1(bool b) {
        var arr = b ? ["1"] : new[] { "2" };
        arr[0] = null; // missing warning
        return arr;
    }

    public string[] M2(bool b) {
        var arr = new[] { "2" };
        arr = b ? ["1"] : arr;
        arr[0] = null; // warning CS8625: Cannot convert null literal to non-nullable reference type.
        return arr;
    }
}
```

--- a/src/Compilers/CSharp/Portable/Binder/Semantics/BestTypeInferrer.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Semantics/BestTypeInferrer.cs
@@ -69,6 +69,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             HashSet<TypeSymbol> candidateTypes = new HashSet<TypeSymbol>(comparer);
             foreach (BoundExpression expr in exprs)
             {
+                Debug.Assert(!NullableWalker.IsTargetTypedExpression(expr));
                 TypeSymbol? type = expr.GetTypeOrFunctionType();
 
                 if (type is { })
@@ -132,9 +133,9 @@ namespace Microsoft.CodeAnalysis.CSharp
             try
             {
                 var conversionsWithoutNullability = conversions.WithNullability(false);
-                TypeSymbol? type1 = expr1.Type;
 
-                if (type1 is { })
+                Debug.Assert(!NullableWalker.IsTargetTypedExpression(expr1));
+                if (expr1.Type is { } type1)
                 {
                     if (type1.IsErrorType())
                     {
@@ -148,9 +149,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                     }
                 }
 
-                TypeSymbol? type2 = expr2.Type;
-
-                if (type2 is { })
+                Debug.Assert(!NullableWalker.IsTargetTypedExpression(expr1));
+                if (expr2.Type is { } type2)
                 {
                     if (type2.IsErrorType())
                     {

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -9317,10 +9317,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             while (conversionOpt != null && conversionOpt != convertedNode)
             {
                 Debug.Assert(conversionOpt.ConversionGroupOpt == conversionGroup);
-                if (!visitResult.LValueType.Type.Equals(conversionOpt.Type, TypeCompareKind.AllNullableIgnoreOptions))
-                {
-                    // visitResult = withType(visitResult, conversionOpt.Type);
-                }
+                Debug.Assert(resultType.Type.Equals.Equals(conversionOpt.Type, TypeCompareKind.AllNullableIgnoreOptions));
                 SetAnalyzedNullability(conversionOpt, visitResult);
                 conversionOpt = conversionOpt.Operand as BoundConversion;
             }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -9317,7 +9317,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             while (conversionOpt != null && conversionOpt != convertedNode)
             {
                 Debug.Assert(conversionOpt.ConversionGroupOpt == conversionGroup);
-                Debug.Assert(resultType.Type.Equals.Equals(conversionOpt.Type, TypeCompareKind.AllNullableIgnoreOptions));
+                Debug.Assert(resultType.Type.Equals(conversionOpt.Type, TypeCompareKind.AllNullableIgnoreOptions));
                 SetAnalyzedNullability(conversionOpt, visitResult);
                 conversionOpt = conversionOpt.Operand as BoundConversion;
             }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -3702,7 +3702,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
                 // Record the final state
                 NullableFlowState resultState = getResultState(node, collectionKind);
-                var resultTypeWithState = TypeWithState.Create(node.Type, resultState);
+                var resultTypeWithState = TypeWithState.Create(strippedTargetCollectionType, resultState);
                 SetAnalyzedNullability(node, resultTypeWithState);
                 return resultTypeWithState;
             }
@@ -4570,7 +4570,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// This is done using <see cref="TargetTypedAnalysisCompletion"/>. All registered completions must be processed
         /// (ie. analyzed via some conversion) before the nullable analysis completes.
         /// </summary>
-        private static bool IsTargetTypedExpression(BoundExpression node)
+        internal static bool IsTargetTypedExpression(BoundExpression node)
         {
             return node is BoundConditionalOperator { WasTargetTyped: true } or
                            BoundConvertedSwitchExpression { WasTargetTyped: true } or
@@ -4613,7 +4613,11 @@ namespace Microsoft.CodeAnalysis.CSharp
             {
                 var (returnExpr, resultType, _) = returns[i];
                 resultTypes.Add(resultType);
-                placeholdersBuilder.Add(CreatePlaceholderIfNecessary(returnExpr, resultType));
+
+                if (!IsTargetTypedExpression(returnExpr))
+                {
+                    placeholdersBuilder.Add(CreatePlaceholderIfNecessary(returnExpr, resultType));
+                }
             }
 
             var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
@@ -5816,6 +5820,14 @@ namespace Microsoft.CodeAnalysis.CSharp
             if (wasTargetTyped)
             {
                 resultType = null;
+            }
+            else if (IsTargetTypedExpression(consequence))
+            {
+                resultType = alternativeRValue.Type;
+            }
+            else if (IsTargetTypedExpression(alternative))
+            {
+                resultType = consequenceRValue.Type;
             }
             else
             {
@@ -9305,7 +9317,11 @@ namespace Microsoft.CodeAnalysis.CSharp
             while (conversionOpt != null && conversionOpt != convertedNode)
             {
                 Debug.Assert(conversionOpt.ConversionGroupOpt == conversionGroup);
-                visitResult = withType(visitResult, conversionOpt.Type);
+                // TODO2: what does it mean for these types to differ? Why is it correct to copy over the top-level nullability when the types have differences beyond nullability?
+                if (!visitResult.LValueType.Type.Equals(conversionOpt.Type, TypeCompareKind.AllNullableIgnoreOptions))
+                {
+                    visitResult = withType(visitResult, conversionOpt.Type);
+                }
                 SetAnalyzedNullability(conversionOpt, visitResult);
                 conversionOpt = conversionOpt.Operand as BoundConversion;
             }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -9317,7 +9317,6 @@ namespace Microsoft.CodeAnalysis.CSharp
             while (conversionOpt != null && conversionOpt != convertedNode)
             {
                 Debug.Assert(conversionOpt.ConversionGroupOpt == conversionGroup);
-                // TODO2: what does it mean for these types to differ? Why is it correct to copy over the top-level nullability when the types have differences beyond nullability?
                 if (!visitResult.LValueType.Type.Equals(conversionOpt.Type, TypeCompareKind.AllNullableIgnoreOptions))
                 {
                     visitResult = withType(visitResult, conversionOpt.Type);

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -9319,15 +9319,15 @@ namespace Microsoft.CodeAnalysis.CSharp
                 Debug.Assert(conversionOpt.ConversionGroupOpt == conversionGroup);
                 if (!visitResult.LValueType.Type.Equals(conversionOpt.Type, TypeCompareKind.AllNullableIgnoreOptions))
                 {
-                    visitResult = withType(visitResult, conversionOpt.Type);
+                    // visitResult = withType(visitResult, conversionOpt.Type);
                 }
                 SetAnalyzedNullability(conversionOpt, visitResult);
                 conversionOpt = conversionOpt.Operand as BoundConversion;
             }
 
-            static VisitResult withType(VisitResult visitResult, TypeSymbol newType) =>
-                new VisitResult(TypeWithState.Create(newType, visitResult.RValueType.State),
-                                TypeWithAnnotations.Create(newType, visitResult.LValueType.NullableAnnotation));
+            // static VisitResult withType(VisitResult visitResult, TypeSymbol newType) =>
+            //     new VisitResult(TypeWithState.Create(newType, visitResult.RValueType.State),
+            //                     TypeWithAnnotations.Create(newType, visitResult.LValueType.NullableAnnotation));
         }
 
         /// <summary>

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -9317,7 +9317,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             while (conversionOpt != null && conversionOpt != convertedNode)
             {
                 Debug.Assert(conversionOpt.ConversionGroupOpt == conversionGroup);
-                Debug.Assert(resultType.Type.Equals(conversionOpt.Type, TypeCompareKind.AllNullableIgnoreOptions));
+                Debug.Assert(conversionOpt.Type.Equals(resultType.Type, TypeCompareKind.AllNullableIgnoreOptions));
                 SetAnalyzedNullability(conversionOpt, visitResult);
                 conversionOpt = conversionOpt.Operand as BoundConversion;
             }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -4613,11 +4613,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             {
                 var (returnExpr, resultType, _) = returns[i];
                 resultTypes.Add(resultType);
-
-                if (!IsTargetTypedExpression(returnExpr))
-                {
-                    placeholdersBuilder.Add(CreatePlaceholderIfNecessary(returnExpr, resultType));
-                }
+                placeholdersBuilder.Add(CreatePlaceholderIfNecessary(returnExpr, resultType));
             }
 
             var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -9317,14 +9317,17 @@ namespace Microsoft.CodeAnalysis.CSharp
             while (conversionOpt != null && conversionOpt != convertedNode)
             {
                 Debug.Assert(conversionOpt.ConversionGroupOpt == conversionGroup);
-                Debug.Assert(conversionOpt.Type.Equals(resultType.Type, TypeCompareKind.AllNullableIgnoreOptions));
+
+                // https://github.com/dotnet/roslyn/issues/35046
+                // SetAnalyzedNullability will drop the type if the visitResult.RValueType.Type differs from conversionOpt.Type.
+                // (It will use the top-level nullability from visitResult, though.)
+                //
+                // Here, the visitResult represents the result of visiting the operand.
+                // Ideally, we would use the visitResult to reinfer the types of the containing conversions, and store those results here.
                 SetAnalyzedNullability(conversionOpt, visitResult);
+
                 conversionOpt = conversionOpt.Operand as BoundConversion;
             }
-
-            // static VisitResult withType(VisitResult visitResult, TypeSymbol newType) =>
-            //     new VisitResult(TypeWithState.Create(newType, visitResult.RValueType.State),
-            //                     TypeWithAnnotations.Create(newType, visitResult.LValueType.NullableAnnotation));
         }
 
         /// <summary>

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs
@@ -898,8 +898,11 @@ namespace Microsoft.CodeAnalysis.CSharp
                 resultTypes.Add(armType);
                 Join(ref endState, ref this.State);
 
-                // Build placeholders for inference in order to preserve annotations.
-                placeholderBuilder.Add(CreatePlaceholderIfNecessary(expression, armType.ToTypeWithAnnotations(compilation)));
+                if (!IsTargetTypedExpression(expression))
+                {
+                    // Build placeholders for inference in order to preserve annotations.
+                    placeholderBuilder.Add(CreatePlaceholderIfNecessary(expression, armType.ToTypeWithAnnotations(compilation)));
+                }
             }
 
             SetState(endState);

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
@@ -5501,7 +5501,22 @@ class C
                 """;
 
             var comp = CreateCompilation(source);
-            comp.VerifyEmitDiagnostics();
+            comp.VerifyEmitDiagnostics(
+                // (9,16): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         x[0] = null; // 1
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(9, 16),
+                // (21,16): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         x[0] = null; // 3
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(21, 16),
+                // (27,16): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         x[0] = null; // 4
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(27, 16),
+                // (35,16): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         x[0] = null; // 5
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(35, 16),
+                // (41,16): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         x[0] = null; // 6
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(41, 16));
         }
     }
 }

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
@@ -5386,6 +5386,64 @@ class C
             }
         }
 
-        // TODO2: test collection-exprs in lambda returns to exercise all BestTypeInferrer usages in NullableWalker
+        [Fact]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/74609")]
+        public void CollectionExpression_NestedNullability_06()
+        {
+            var source = """
+                #nullable enable
+
+                public class C
+                {
+                    public string[] M1(bool b)
+                    {
+                        var lam = () =>
+                        {
+                            if (b)
+                                return ["1"];
+
+                            return new[] { "2" };
+                        };
+
+                        var arr = lam();
+                        arr[0] = null; // 1
+                        return arr;
+                    }
+
+                    public string[] M2(bool b)
+                    {
+                        var lam = () => new[] { "2" };
+                        var arr = lam();
+                        arr = b ? ["1"] : arr;
+                        arr[0] = null; // 2
+                        return arr;
+                    }
+                }
+                """;
+
+            var comp = CreateCompilation(source);
+            // The expected nullable warnings are not reported here.
+            // https://github.com/dotnet/roslyn/issues/74609
+            comp.VerifyEmitDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var root = tree.GetRoot();
+            var collectionExprs = root.DescendantNodes().OfType<CollectionExpressionSyntax>().ToArray();
+            Assert.Equal(2, collectionExprs.Length);
+            foreach (var collectionExpr in collectionExprs)
+            {
+                var typeInfo = model.GetTypeInfo(collectionExpr);
+                var type = (IArrayTypeSymbol)typeInfo.ConvertedType;
+                Assert.Equal("System.String[]", type.ToTestDisplayString());
+                Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.NullableAnnotation);
+
+                // The arrays have unexpected oblivious element nullability.
+                // https://github.com/dotnet/roslyn/issues/74609
+                Assert.Equal(PublicNullableAnnotation.None, type.ElementNullableAnnotation);
+            }
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
@@ -5261,6 +5261,29 @@ class C
         }
 
         [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
+        public void CollectionExpression_NestedNullability_01_RhsTargetTyped()
+        {
+            var source = """
+                #nullable enable
+
+                var b = false;
+                var arr = b ? new[] { "1" } : ["2"];
+                """;
+
+            var comp = CreateCompilation(source);
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var root = tree.GetRoot();
+            var collectionExpr = root.DescendantNodes().OfType<CollectionExpressionSyntax>().Single();
+            var typeInfo = model.GetTypeInfo(collectionExpr);
+            var type = (IArrayTypeSymbol)typeInfo.ConvertedType;
+            Assert.Equal("System.String[]", type.ToTestDisplayString());
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.NullableAnnotation);
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.ElementNullableAnnotation);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
         public void CollectionExpression_NestedNullability_02()
         {
             var source = """

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
@@ -14,6 +14,7 @@ using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Test.Utilities;
 using Roslyn.Test.Utilities;
+using Roslyn.Utilities;
 using Xunit;
 using PublicNullableAnnotation = Microsoft.CodeAnalysis.NullableAnnotation;
 using PublicNullableFlowState = Microsoft.CodeAnalysis.NullableFlowState;
@@ -5235,5 +5236,156 @@ class C
                 Assert.Equal(PublicNullableAnnotation.None, typeInfo.Type.NullableAnnotation);
             }
         }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
+        public void CollectionExpression_NestedNullability_01()
+        {
+            var source = """
+                #nullable enable
+
+                var b = false;
+                var arr = b ? ["1"] : new[] { "2" };
+                """;
+
+            var comp = CreateCompilation(source);
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var root = tree.GetRoot();
+            var collectionExpr = root.DescendantNodes().OfType<CollectionExpressionSyntax>().Single();
+            var typeInfo = model.GetTypeInfo(collectionExpr);
+            var type = (IArrayTypeSymbol)typeInfo.ConvertedType;
+            Assert.Equal("System.String[]", type.ToTestDisplayString());
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.NullableAnnotation);
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.ElementNullableAnnotation);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
+        public void CollectionExpression_NestedNullability_02()
+        {
+            var source = """
+                #nullable enable
+
+                using System;
+                using System.Collections.Generic;
+                using System.Linq.Expressions;
+
+                class C
+                {
+                    void M()
+                    {
+                        string[] x = ["1"];
+                    }
+                }
+                """;
+
+            var comp = CreateCompilation(source);
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var root = tree.GetRoot();
+            var collectionExpr = root.DescendantNodes().OfType<CollectionExpressionSyntax>().Single();
+            var typeInfo = model.GetTypeInfo(collectionExpr);
+            var type = (IArrayTypeSymbol)typeInfo.ConvertedType;
+            Assert.Equal("System.String[]", type.ToTestDisplayString());
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.NullableAnnotation);
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.ElementNullableAnnotation);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
+        public void CollectionExpression_NestedNullability_03()
+        {
+            var source = """
+                #nullable enable
+
+                var b = false;
+                var arr = b switch { true => ["1"], false => new[] { "2" } };
+                """;
+
+            var comp = CreateCompilation(source);
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var root = tree.GetRoot();
+            var collectionExpr = root.DescendantNodes().OfType<CollectionExpressionSyntax>().Single();
+            var typeInfo = model.GetTypeInfo(collectionExpr);
+            var type = (IArrayTypeSymbol)typeInfo.ConvertedType;
+            Assert.Equal("System.String[]", type.ToTestDisplayString());
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.NullableAnnotation);
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.ElementNullableAnnotation);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
+        public void CollectionExpression_NestedNullability_04()
+        {
+            var source = """
+                #nullable enable
+
+                var arr = new[] { ["1"], new[] { "2" } };
+                """;
+
+            var comp = CreateCompilation(source);
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var root = tree.GetRoot();
+            var collectionExpr = root.DescendantNodes().OfType<CollectionExpressionSyntax>().Single();
+            var typeInfo = model.GetTypeInfo(collectionExpr);
+            var type = (IArrayTypeSymbol)typeInfo.ConvertedType;
+            Assert.Equal("System.String[]", type.ToTestDisplayString());
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.NullableAnnotation);
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.ElementNullableAnnotation);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
+        public void CollectionExpression_NestedNullability_05()
+        {
+            var source = """
+                #nullable enable
+                public class C
+                {
+                    public string[] M1(bool b)
+                    {
+                        var arr = b ? ["1"] : new[] { "2" };
+                        arr[0] = null; // 1
+                        return arr;
+                    }
+
+                    public string[] M2(bool b)
+                    {
+                        var arr = new[] { "2" };
+                        arr = b ? ["1"] : arr;
+                        arr[0] = null; // 2
+                        return arr;
+                    }
+                }
+                """;
+
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (7,18): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         arr[0] = null; // 1
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(7, 18),
+                // (15,18): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         arr[0] = null; // 2
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(15, 18));
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var root = tree.GetRoot();
+            var collectionExprs = root.DescendantNodes().OfType<CollectionExpressionSyntax>().ToArray();
+            Assert.Equal(2, collectionExprs.Length);
+            foreach (var collectionExpr in collectionExprs)
+            {
+                var typeInfo = model.GetTypeInfo(collectionExpr);
+                var type = (IArrayTypeSymbol)typeInfo.ConvertedType;
+                Assert.Equal("System.String[]", type.ToTestDisplayString());
+                Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.NullableAnnotation);
+                Assert.Equal(PublicNullableAnnotation.NotAnnotated, type.ElementNullableAnnotation);
+            }
+        }
+
+        // TODO2: test collection-exprs in lambda returns to exercise all BestTypeInferrer usages in NullableWalker
     }
 }

